pbjson-build/src/generator/message.rs

```diff
@@ -753,151 +753,207 @@ fn write_deserialize_field<W: Write>(
+                        resolver.rust_type(&one_of.path),
+                        field.rust_type_name()
+                    )?;
+                }
+                None => {
+                    write!(
+                        writer,
+                        "map.next_value::<::std::option::Option<_>>()?.map({}::{})",
+                        resolver.rust_type(&one_of.path),
+                        field.rust_type_name()
+                    )?;
+                }
+            },
+            FieldType::Enum(path) => {
                 write!(
                     writer,
-                    "map.next_value::<{}>()? as i32",
-                    resolver.rust_type(path)
+                    "map.next_value::<::std::option::Option<{}>>()?.map(|x| {}::{}(x as i32))",
+                    resolver.rust_type(path),
+                    resolver.rust_type(&one_of.path),
+                    field.rust_type_name()
                 )?;
             }
+            FieldType::Message(_) => writeln!(
+                writer,
+                "map.next_value::<::std::option::Option<_>>()?.map({}::{})",
+                resolver.rust_type(&one_of.path),
+                field.rust_type_name()
+            )?,
+            FieldType::Map(_, _) => unreachable!("one of cannot contain map fields"),
         },
-        FieldType::Map(key, value) => {
-            writeln!(writer)?;
-            match btree_map {
-                true => write!(
-                    writer,
-                    "{}map.next_value::<std::collections::BTreeMap<",
-                    Indent(indent + 2),
-                )?,
-                false => write!(
-                    writer,
-                    "{}map.next_value::<std::collections::HashMap<",
-                    Indent(indent + 2),
-                )?,
-            }
 
-            let map_k = match key {
-                ScalarType::Bytes | ScalarType::F32 | ScalarType::F64 => {
-                    panic!("protobuf disallows maps with floating point or bytes keys")
+        None => match &field.field_type {
+            FieldType::Scalar(scalar) => {
+                write_encode_scalar_field(indent + 1, *scalar, field.field_modifier, writer)?;
+            }
+            FieldType::Enum(path) => match field.field_modifier {
+                FieldModifier::Optional => {
+                    write!(
+                        writer,
+                        "map.next_value::<::std::option::Option<{}>>()?.map(|x| x as i32)",
+                        resolver.rust_type(path)
+                    )?;
                 }
-                _ if key.is_numeric() => {
+                FieldModifier::Repeated => {
                     write!(
                         writer,
-                        "::pbjson::private::NumberDeserialize<{}>",
-                        key.rust_type()
+                        "Some(map.next_value::<Vec<{}>>()?.into_iter().map(|x| x as i32).collect())",
```

Comment:
I really got confused about `Vec<{}>` and it took me a while to realize that `{}` is NOT Rust code, but placeholder for `write!`. I've said this below, but I think checking in the generated test code would help a lot.

Reply:
Also: you're really careful to use fully-qualified names for all types including `Option`. I wonder if this should also be done for `Vec` (`::std::vec::Vec`)?

Reply:
Filed #80